File: src/fairshare/settlement.py
"""Greedy min-cash-flow settlement algorithm."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Payment:
    """A single settlement payment: *frm* pays *amount_cents* to *to*."""

    frm: str
    to: str
    amount_cents: int


def settle(balances: dict[str, int]) -> list[Payment]:
    """Compute the minimum cash-flow settlement using a greedy algorithm.

    Always pairs the largest debtor with the largest creditor.
    Note: this is a greedy heuristic and may not be globally optimal
    for all balance distributions, but it is fast and produces a small
    number of transactions in practice.

    Balances with zero value are ignored.
    """
    # Work with mutable copies, filtering out zeros
    net = {p: b for p, b in balances.items() if b != 0}
    payments: list[Payment] = []

    while True:
        debtors = sorted(((b, p) for p, b in net.items() if b < 0), key=lambda x: x[0])
        creditors = sorted(((b, p) for p, b in net.items() if b > 0), key=lambda x: -x[0])

        if not debtors or not creditors:
            break

        debt_amt, debtor = debtors[0]
        cred_amt, creditor = creditors[0]

        amount = min(-debt_amt, cred_amt)
        payments.append(Payment(frm=debtor, to=creditor, amount_cents=amount))

        net[debtor] += amount
        net[creditor] -= amount

        if net[debtor] == 0:
            del net[debtor]
        if net[creditor] == 0:
            del net[creditor]

    return payments

File: web/src/domain/settlement.ts
export type Payment = {
  readonly frm: string;
  readonly to: string;
  readonly amount_cents: number;
};

export function settle(balances: Record<string, number>): Payment[] {
  const net: Record<string, number> = {};
  for (const [person, amount] of Object.entries(balances)) {
    if (amount !== 0) {
      net[person] = amount;
    }
  }

  const payments: Payment[] = [];

  while (true) {
    const debtors = Object.entries(net)
      .filter(([, b]) => b < 0)
      .sort((a, b) => a[1] - b[1] || a[0].localeCompare(b[0]));
    const creditors = Object.entries(net)
      .filter(([, b]) => b > 0)
      .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));

    if (debtors.length === 0 || creditors.length === 0) {
      break;
    }

    const [debtor, debtAmt] = debtors[0];
    const [creditor, credAmt] = creditors[0];
    const amount = Math.min(-debtAmt, credAmt);
    payments.push({ frm: debtor, to: creditor, amount_cents: amount });

    net[debtor] += amount;
    net[creditor] -= amount;

    if (net[debtor] === 0) {
      delete net[debtor];
    }
    if (net[creditor] === 0) {
      delete net[creditor];
    }
  }

  return payments;
}
